File: src/main/java/io/github/funofprograming/context/impl/ApplicationContextImpl.java
package io.github.funofprograming.context.impl;

import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Optional;
import java.util.Set;

import io.github.funofprograming.context.ApplicationContext;
import io.github.funofprograming.context.ApplicationContextMergeStrategy;
import io.github.funofprograming.context.Key;
import lombok.AccessLevel;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;

/**
 * Plain vanilla implementation of {@linkplain ApplicationContext}
 * 
 * 
 * @author Akshay Jain
 *
 */
@EqualsAndHashCode
@ToString(of = {"name"})
public class ApplicationContextImpl implements ApplicationContext {

    @Getter
    private final String name;
    
    @Getter
    private final Set<Key<?>> permittedKeys;
    
    @Getter(AccessLevel.PROTECTED)
    private final Map<Key<?>, Optional<Object>> store; 
    
    /**
     * Initialize context with a name and initialize store as a {@linkplain HashMap}
     * 
     * @param name
     */
    public ApplicationContextImpl(String name) 
    {
        this(name, null);
    }
    
    /**
     * Initialize context with a name, set of permittedKeys and initialize store as a {@linkplain HashMap} 
     * 
     * @param name
     */
    public ApplicationContextImpl(String name, Set<Key<?>> permittedKeys) 
    {
        this.name = name;
        this.permittedKeys = Optional.ofNullable(permittedKeys).map(p->Collections.unmodifiableSet(p)).orElse(Collections.emptySet());
        this.store = new LinkedHashMap<>();
    }
    
    /**
     * Validate a give key if this context was initialized with a non-empty set of permittedKeys
     * 
     * @param key
     */
    protected void validateKey(Key<?> key)
    {
        if(Optional.ofNullable(getPermittedKeys()).map(p->!p.isEmpty() && !p.contains(key)).orElse(Boolean.FALSE))
        {
            throw new InvalidKeyException("Invalid key:"+key+". Valid keys for this context are: "+getPermittedKeys());
        }
    }
    
    /**
     * {@inheritDoc}
     */
    public boolean isKeyValid(Key<?> key)
    {
        try 
        {
            validateKey(key);
            return true;
        } 
        catch(Throwable t)
        {
            return false;
        }
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> void addIfNotPresent(Key<T> key, T value)
    {
        validateKey(key);
        
        getStore().putIfAbsent(key, Optional.ofNullable(value));
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings("unchecked")
    @Override
    public <T> T addWithOverwrite(Key<T> key, T value) 
    {
        validateKey(key);
        
        return (T)getStore().put(key, Optional.ofNullable(value));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> void add(Key<T> key, T value)
    {
        validateKey(key);
        
        getStore().put(key, Optional.ofNullable(value));
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public <T> boolean exists(Key<T> key) 
    {
        validateKey(key);
        
        return getStore().containsKey(key);
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings("unchecked")
    @Override
    public <T> T fetch(Key<T> key) 
    {
        validateKey(key);
        
        return Optional.ofNullable(getStore().get(key)).flatMap(o->o.map(ob->(T)ob)).orElse(null);
    }

    /**
     * {@inheritDoc}
     */
    @SuppressWarnings("unchecked")
    @Override
    public <T> T erase(Key<T> key)
    {
        validateKey(key);
        
        return Optional.ofNullable(getStore().remove(key)).flatMap(o->o.map(ob->(T)ob)).orElse(null);
    }
    
    /**
     * {@inheritDoc}
     */
    @Override
    public void clear() 
    {
        getStore().clear();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public Set<Key<?>> keySet()
    {
        return Collections.unmodifiableSet(getStore().keySet());
    } 

    @SuppressWarnings({ "rawtypes", "unchecked" })
    @Override
    public void merge(ApplicationContext other, ApplicationContextMergeStrategy mergeStrategy)
    {
        Set<Key<?>> keysOther = other.keySet();
        
        for(Key keyOther: keysOther)
        {
            if(!isKeyValid(keyOther))
            {
                continue;
            }
            
            Object newValue = other.fetch(keyOther);
            
            if(this.exists(keyOther)) 
            {
                Object value = mergeStrategy.merge(keyOther, this.fetch(keyOther), newValue);
                addWithOverwrite(keyOther, value);
            }
            else
            {
                addIfNotPresent(keyOther, newValue);
            }
        }
    }
}

File: src/main/kotlin/io/github/funofprograming/context/impl/ApplicationContextImpl.kt
package io.github.funofprograming.context.impl

import io.github.funofprograming.context.ApplicationContext
import io.github.funofprograming.context.ApplicationContextMergeStrategy
import io.github.funofprograming.context.Key

/**
 * Plain vanilla implementation of {@linkplain ApplicationContext}
 *
 *
 * @author Akshay Jain
 * @property name name of the context
 * @property permittedKeys optional set of permitted keys for this context useful when restricted key set to be allowed. null or empty means any key allowed
 */
open class ApplicationContextImpl(private val name: String, private val permittedKeys: Set<Key<*>>? = null) : ApplicationContext {

    private val store: MutableMap<Key<*>, Any?> = mutableMapOf()

    public constructor(name: String, permittedKeys: Set<Key<*>>? = null, initialCopy: ApplicationContext) : this(name, permittedKeys) {
        merge(initialCopy, ApplicationContextMergeOverwriteStrategy(permittedKeys))
    }

    private fun validateKey(key: Key<*>) {
        if(!permittedKeys.isNullOrEmpty() && !permittedKeys.contains(key))
            throw InvalidKeyException("Invalid key: $key. Valid keys for this context are: $permittedKeys")
    }

    override fun getName(): String {
        return name
    }

    override fun getPermittedKeys(): Set<Key<*>>? {
        return permittedKeys
    }

    override fun isKeyValid(key: Key<*>): Boolean {
        try{
            validateKey(key)
            return true
        } catch (e:InvalidKeyException) {
            return false
        }
    }

    override fun <T> addIfNotPresent(key: Key<T>, value: T?) {
        validateKey(key)
        store.putIfAbsent(key, value)
    }

    override fun <T> addWithOverwrite(key: Key<T>, value: T?): Any? {
        validateKey(key)
        return store.put(key, value)
    }

    override fun <T> add(key: Key<T>, value: T?) {
        addWithOverwrite(key, value)
    }

    override fun <T> exists(key: Key<T>): Boolean {
        validateKey(key);
        return store.contains(key)
    }

    override fun <T> fetch(key: Key<T>): T? {
        validateKey(key);
        return store.get(key) as T
    }

    override fun <T> erase(key: Key<T>): T? {
        validateKey(key)
        return store.remove(key) as T
    }

    override fun clear() = store.clear()

    override fun keySet(): Set<Key<*>>? = store.keys.toSet()

    override fun merge(other: ApplicationContext?, mergeStrategy: ApplicationContextMergeStrategy) {

        val keysOther:Set<Key<Any>>? = other?.keySet() as Set<Key<Any>>

        for (keyOther in keysOther?.toSet()?:emptySet()) {
            if (!isKeyValid(keyOther)) {
                continue
            }

            val newValue = other.fetch(keyOther)

            if (exists(keyOther)) {
                val value = mergeStrategy.merge(keyOther, fetch(keyOther), newValue)
                addWithOverwrite(keyOther, value)
            } else {
                addIfNotPresent(keyOther, newValue)
            }
        }
    }

    override fun clone(cloneName: String, clonePermittedKeys: Set<Key<*>>?): ApplicationContext = ApplicationContextImpl(cloneName, clonePermittedKeys, this)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as ApplicationContextImpl

        return name == other.name
    }

    override fun hashCode(): Int {
        return name.hashCode()
    }
}